File: experimental/s3-client-annotation-processor/src/test/java/ru/tinkoff/kora/s3/client/annotation/processor/S3PutTest.java
package ru.tinkoff.kora.s3.client.annotation.processor;


import org.junit.jupiter.api.Test;
import ru.tinkoff.kora.s3.client.$S3ClientConfig_UploadConfig_ConfigValueExtractor;
import ru.tinkoff.kora.s3.client.S3Client;
import ru.tinkoff.kora.s3.client.model.response.UploadedPart;

import java.io.ByteArrayInputStream;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.ThreadLocalRandom;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;

class S3PutTest extends AbstractS3ClientTest {
    @Test
    public void testPutByteArray() {
        var client = this.compile("""
            @S3.Client
            public interface Client {
                @S3.Put
                String put(@S3.Bucket String bucket, String key, byte[] data);
            }
            """);

        var bytes = UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8);

        when(s3Client.putObject(any(), eq("bucket"), eq("key"), any(), same(bytes), eq(0), eq(bytes.length))).thenReturn("etag");

        assertThat(client.<String>invoke("put", "bucket", "key", bytes)).isEqualTo("etag");

        verify(s3Client).putObject(any(), eq("bucket"), eq("key"), any(), same(bytes), eq(0), eq(bytes.length));
        reset(s3Client);
    }

    @Test
    public void testPutByteBuffer() {
        var client = this.compile("""
            @S3.Client
            public interface Client {
                @S3.Put
                String put(@S3.Bucket String bucket, String key, ByteBuffer data);
            }
            """);

        var bytes = UUID.randomUUID().toString().getBytes(StandardCharsets.UTF_8);

        when(s3Client.putObject(any(), eq("bucket"), eq("key"), any(), same(bytes), eq(0), eq(bytes.length - 1))).thenReturn("etag");

        assertThat(client.<String>invoke("put", "bucket", "key", ByteBuffer.wrap(bytes).slice(0, bytes.length - 1))).isEqualTo("etag");

        verify(s3Client).putObject(any(), eq("bucket"), eq("key"), any(), same(bytes), eq(0), eq(bytes.length - 1));
        reset(s3Client);
    }

    @Test
    public void testPutWithContentWriter() throws Exception {
        var client = this.compile("""
            @S3.Client
            public interface Client {
                @S3.Put
                String put(@S3.Bucket String bucket, String key, S3Client.ContentWriter writer);
            }
            """);

        var writer = mock(S3Client.ContentWriter.class);
        when(writer.length()).thenReturn(5L);

        when(s3Client.putObject(any(), eq("bucket"), eq("key"), any(), same(writer))).thenReturn("etag-content");

        assertThat(client.<String>invoke("put", "bucket", "key", writer)).isEqualTo("etag-content");

        verify(s3Client).putObject(any(), eq("bucket"), eq("key"), any(), same(writer));
        reset(s3Client);
    }

    @Test
    public void testPutInputStream() throws Exception {
        var client = this.compile("""
            @S3.Client
            public interface Client {
                @S3.Put
                String put(@S3.Bucket String bucket, String key, InputStream writer);
            }
            """);

        var bytes = new byte[8 * 1024 * 1024];
        ThreadLocalRandom.current().nextBytes(bytes);
        var is = new ByteArrayInputStream(bytes);
        var part1 = new UploadedPart(null, null, null, null, null, "etag1", 1, 5 * 1024 * 1024);
        var part2 = new UploadedPart(null, null, null, null, null, "etag2", 2, 3 * 1024 * 1024);

        when(s3Client.createMultipartUpload(any(), eq("bucket"), eq("key"), any())).thenReturn("multipartid");
        when(s3Client.uploadPart(any(), eq("bucket"), eq("key"), eq("multipartid"), eq(1), any(), eq(0), eq(5 * 1024 * 1024))).thenReturn(part1);
        when(s3Client.uploadPart(any(), eq("bucket"), eq("key"), eq("multipartid"), eq(2), any(), eq(0), eq(3 * 1024 * 1024))).thenReturn(part2);
        when(s3Client.completeMultipartUpload(any(), eq("bucket"), eq("key"), eq("multipartid"), eq(List.of(part1, part2)), any())).thenReturn("etag-final");
        when(config.upload()).thenReturn(new $S3ClientConfig_UploadConfig_ConfigValueExtractor.UploadConfig_Defaults());

        assertThat(client.<String>invoke("put", "bucket", "key", is)).isEqualTo("etag-final");

        verify(s3Client).createMultipartUpload(any(), eq("bucket"), eq("key"), any());
        verify(s3Client).uploadPart(any(), eq("bucket"), eq("key"), eq("multipartid"), eq(1), any(), eq(0), eq(5 * 1024 * 1024));
        verify(s3Client).uploadPart(any(), eq("bucket"), eq("key"), eq("multipartid"), eq(2), any(), eq(0), eq(3 * 1024 * 1024));
        verify(s3Client).completeMultipartUpload(any(), eq("bucket"), eq("key"), eq("multipartid"), eq(List.of(part1, part2)), any());
        reset(s3Client);
    }
}

File: aws/aws-s3-symbol-processor/src/test/kotlin/ru/tinkoff/kora/aws/s3/symbol/processor/S3PutTest.kt
package ru.tinkoff.kora.aws.s3.symbol.processor

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.*
import org.mockito.Mockito.*
import ru.tinkoff.kora.aws.s3.`$S3ClientConfig_UploadConfig_ConfigValueExtractor`
import ru.tinkoff.kora.aws.s3.S3Client
import ru.tinkoff.kora.aws.s3.S3Client.ContentWriter
import ru.tinkoff.kora.aws.s3.model.response.UploadedPart
import java.io.ByteArrayInputStream
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.concurrent.ThreadLocalRandom


internal class S3PutTest : AbstractS3ClientTest() {
    @Test
    fun testPutByteArray() {
        val client = this.compile(
            """
            @S3.Client
            interface Client {
                @S3.Put
                fun put(@S3.Bucket bucket: String, key: String,  data: ByteArray): String
            }
            
            """.trimIndent()
        )

        val bytes = UUID.randomUUID().toString().toByteArray(StandardCharsets.UTF_8)

        `when`(
            s3Client.putObject(
                any(),
                eq("bucket"),
                eq("key"),
                any(),
                same(bytes),
                eq(0),
                eq(bytes.size)
            )
        ).thenReturn("etag")

        assertThat(client.invoke<String?>("put", "bucket", "key", bytes)).isEqualTo("etag")

        verify(s3Client).putObject(
            any(),
            eq("bucket"),
            eq("key"),
            any(),
            same(bytes),
            eq(0),
            eq(bytes.size)
        )
        reset(s3Client)
    }

    @Test
    fun testPutByteBuffer() {
        val client = this.compile(
            """
            @S3.Client
            interface Client {
                @S3.Put
                fun put(@S3.Bucket bucket: String, key: String, data: ByteBuffer): String
            }
            
            """.trimIndent()
        )

        val bytes = UUID.randomUUID().toString().toByteArray(StandardCharsets.UTF_8)

        `when`(
            s3Client.putObject(
                any(),
                eq("bucket"),
                eq("key"),
                any(),
                same(bytes),
                eq(0),
                eq(bytes.size - 1)
            )
        ).thenReturn("etag")

        assertThat(client.invoke<String?>("put", "bucket", "key", ByteBuffer.wrap(bytes).slice(0, bytes.size - 1))).isEqualTo("etag")

        verify(s3Client).putObject(
            any(),
            eq("bucket"),
            eq("key"),
            any(),
            same(bytes),
            eq(0),
            eq(bytes.size - 1)
        )
        reset(s3Client)
    }

    @Test
    @Throws(Exception::class)
    fun testPutWithContentWriter() {
        val client = this.compile(
            """
            @S3.Client
            interface Client {
                @S3.Put
                fun put(@S3.Bucket bucket: String, key: String, writer: S3Client.ContentWriter): String
            }
            
            """.trimIndent()
        )

        val writer = mock(ContentWriter::class.java)
        `when`(writer.length()).thenReturn(5L)

        `when`(
            s3Client.putObject(
                any(),
                eq("bucket"),
                eq("key"),
                any(),
                same(writer)
            )
        ).thenReturn("etag-content")

        assertThat(client.invoke<String?>("put", "bucket", "key", writer)).isEqualTo("etag-content")

        verify(s3Client).putObject(
            any(),
            eq("bucket"),
            eq("key"),
            any(),
            same(writer)
        )
        reset<S3Client?>(s3Client)
    }

    @Test
    @Throws(Exception::class)
    fun testPutInputStream() {
        val client = this.compile(
            """
            @S3.Client
            interface Client {
                @S3.Put
                fun put(@S3.Bucket bucket: String, key: String, writer: InputStream): String
            }
            """.trimIndent()
        )

        val bytes = ByteArray(8 * 1024 * 1024)
        ThreadLocalRandom.current().nextBytes(bytes)
        val `is` = ByteArrayInputStream(bytes)
        val part1 = UploadedPart(null, null, null, null, null, "etag1", 1, (5 * 1024 * 1024).toLong())
        val part2 = UploadedPart(null, null, null, null, null, "etag2", 2, (3 * 1024 * 1024).toLong())

        `when`(
            s3Client.createMultipartUpload(
                any(),
                eq("bucket"),
                eq("key"),
                any()
            )
        ).thenReturn("multipartid")
        `when`(
            s3Client.uploadPart(
                any(),
                eq("bucket"),
                eq("key"),
                eq("multipartid"),
                eq(1),
                any(),
                eq(0),
                eq(5 * 1024 * 1024)
            )
        ).thenReturn(part1)
        `when`(
            s3Client.uploadPart(
                any(),
                eq("bucket"),
                eq("key"),
                eq("multipartid"),
                eq(2),
                any(),
                eq(0),
                eq(3 * 1024 * 1024)
            )
        ).thenReturn(part2)
        `when`(
            s3Client.completeMultipartUpload(
                any(),
                eq("bucket"),
                eq("key"),
                eq("multipartid"),
                eq(listOf(part1, part2)),
                any()
            )
        ).thenReturn("etag-final")
        `when`(config.upload()).thenReturn(`$S3ClientConfig_UploadConfig_ConfigValueExtractor`.UploadConfig_Defaults())

        assertThat(client.invoke<String?>("put", "bucket", "key", `is`)).isEqualTo("etag-final")

        verify(s3Client).createMultipartUpload(
            any(),
            eq("bucket"),
            eq("key"),
            any()
        )
        verify(s3Client).uploadPart(
            any(),
            eq("bucket"),
            eq("key"),
            eq("multipartid"),
            eq(1),
            any(),
            eq(0),
            eq(5 * 1024 * 1024)
        )
        verify(s3Client).uploadPart(
            any(),
            eq("bucket"),
            eq("key"),
            eq("multipartid"),
            eq(2),
            any(),
            eq(0),
            eq(3 * 1024 * 1024)
        )
        verify(s3Client).completeMultipartUpload(
            any(),
            eq("bucket"),
            eq("key"),
            eq("multipartid"),
            eq(listOf(part1, part2)),
            any()
        )
        reset(s3Client)
    }
}
